File: be/app/services/novedades_service.py
"""
Servicio de novedades e incidencias.

Permite a los técnicos reportar incidencias sobre entregas, citas y productos.
El administrador revisa, cambia estado y genera compensaciones (cupones).
"""
from datetime import date, datetime
from typing import Optional
from sqlalchemy import select
from sqlalchemy.orm import Session

from app.models.novedad import CuponDescuento, Novedad, NovedadHistorial
from app.models.tecnico import Tecnico
from app.models.user import User


# ── Tipos de novedad ────────────────────────────────────────────────
TIPOS_NOVEDAD = [
    "Retraso en entrega",
    "Cliente ausente",
    "Cliente no recibió el pedido",
    "Dirección incorrecta",
    "Producto dañado",
    "Producto faltante",
    "Producto equivocado",
    "Pérdida de producto",
    "Robo de productos",
    "Accidente",
    "Problema con el vehículo/transporte",
    "Problema técnico",
    "Problema durante una cita",
    "Cita no realizada",
    "Cita reprogramada",
    "Cliente solicita reprogramación",
    "Problema con instalación",
    "Retraso que afecta cita",
    "Otro",
    # Tipos de devolución
    "Producto recibido con daños",
    "Producto incompleto",
    "Cliente no tenía el producto disponible",
    "Producto diferente al solicitado",
    "Devolución no realizada",
    "Problema durante la recogida",
]

ESTADOS_NOVEDAD = ["Pendiente", "En revisión", "Aprobada", "Rechazada", "Resuelta", "Cerrada"]
PRIORIDADES = ["baja", "normal", "alta", "urgente"]


# ── CRUD novidades ──────────────────────────────────────────────────

def crear_novedad(
    db: Session,
    *,
    id_tecnico: int,
    tipo_novedad: str,
    descripcion: str,
    prioridad: str = "normal",
    id_pedido: Optional[int] = None,
    id_cita: Optional[int] = None,
    id_cliente: Optional[int] = None,
    id_devolucion: Optional[int] = None,
    lugar_ocurrencia: Optional[str] = None,
    evidencia_url: Optional[str] = None,
) -> Novedad:
    """Crea una novedad y registra el primer evento en el historial."""
    novedad = Novedad(
        id_tecnico_n=id_tecnico,
        tipo_novedad=tipo_novedad,
        descripcion_novedad=descripcion,
        prioridad=prioridad,
        estado_novedad="Pendiente",
        id_pedido=id_pedido,
        id_cita=id_cita,
        id_cliente=id_cliente,
        id_devolucion=id_devolucion,
        lugar_ocurrencia=lugar_ocurrencia,
        evidencia_url=evidencia_url,
    )
    db.add(novedad)
    db.flush()

    # Historial: creación
    db.add(NovedadHistorial(
        id_novedad=novedad.id_novedad,
        id_usuario=None,
        accion="Creación",
        detalle=f"Novedad reportada: {tipo_novedad}",
    ))
    db.commit()
    db.refresh(novedad)
    return novedad


def listar_novedades_tecnico(db: Session, id_tecnico: int) -> list[Novedad]:
    """Lista novedades reportadas por un técnico específico."""
    return (
        db.query(Novedad)
        .filter(Novedad.id_tecnico_n == id_tecnico)
        .order_by(Novedad.fecha_reporte_novedad.desc())
        .all()
    )


def listar_novedades_admin(
    db: Session,
    *,
    id_pedido: Optional[int] = None,
    id_tecnico: Optional[int] = None,
    id_cliente: Optional[int] = None,
    tipo_novedad: Optional[str] = None,
    estado: Optional[str] = None,
    prioridad: Optional[str] = None,
    fecha_desde: Optional[date] = None,
    fecha_hasta: Optional[date] = None,
) -> list[Novedad]:
    """Lista todas las novedades con filtros opcionales (admin)."""
    q = db.query(Novedad)
    if id_pedido is not None:
        q = q.filter(Novedad.id_pedido == id_pedido)
    if id_tecnico is not None:
        q = q.filter(Novedad.id_tecnico_n == id_tecnico)
    if id_cliente is not None:
        q = q.filter(Novedad.id_cliente == id_cliente)
    if tipo_novedad:
        q = q.filter(Novedad.tipo_novedad == tipo_novedad)
    if estado:
        q = q.filter(Novedad.estado_novedad == estado)
    if prioridad:
        q = q.filter(Novedad.prioridad == prioridad)
    if fecha_desde:
        q = q.filter(Novedad.fecha_reporte_novedad >= datetime.combine(fecha_desde, datetime.min.time()))
    if fecha_hasta:
        q = q.filter(Novedad.fecha_reporte_novedad <= datetime.combine(fecha_hasta, datetime.max.time()))
    return q.order_by(Novedad.fecha_reporte_novedad.desc()).all()


def obtener_novedad(db: Session, id_novedad: int) -> Optional[Novedad]:
    return db.query(Novedad).filter(Novedad.id_novedad == id_novedad).first()


def cambiar_estado_novedad(
    db: Session,
    *,
    id_novedad: int,
    nuevo_estado: str,
    id_admin: int,
    accion_detalle: Optional[str] = None,
) -> Novedad:
    """Cambia el estado de una novedad (solo admin)."""
    novedad = db.query(Novedad).filter(Novedad.id_novedad == id_novedad).first()
    if not novedad:
        raise ValueError("Novedad no encontrada")

    estado_anterior = novedad.estado_novedad
    novedad.estado_novedad = nuevo_estado

    if nuevo_estado in ("Resuelta", "Cerrada"):
        novedad.fecha_resolucion = datetime.now()
        novedad.id_admin_resuelve = id_admin

    # Historial
    db.add(NovedadHistorial(
        id_novedad=id_novedad,
        id_usuario=id_admin,
        accion=f"Cambio de estado: {estado_anterior} → {nuevo_estado}",
        detalle=accion_detalle,
    ))
    db.commit()
    db.refresh(novedad)
    return novedad


def agregar_accion_admin(
    db: Session,
    *,
    id_novedad: int,
    id_admin: int,
    accion: str,
    detalle: Optional[str] = None,
) -> None:
    """Registra una acción del admin en el historial de la novedad."""
    db.add(NovedadHistorial(
        id_novedad=id_novedad,
        id_usuario=id_admin,
        accion=accion,
        detalle=detalle,
    ))
    db.commit()


def responder_novedad(
    db: Session,
    *,
    id_novedad: int,
    id_admin: int,
    respuesta: str,
) -> Novedad:
    """El admin responde a una novedad del técnico."""
    novedad = db.query(Novedad).filter(Novedad.id_novedad == id_novedad).first()
    if not novedad:
        raise ValueError("Novedad no encontrada")

    novedad.accion_admin = respuesta
    novedad.id_admin_resuelve = id_admin
    novedad.estado_novedad = "En revisión"

    db.add(NovedadHistorial(
        id_novedad=id_novedad,
        id_usuario=id_admin,
        accion="Respuesta del administrador",
        detalle=respuesta,
    ))
    db.commit()
    db.refresh(novedad)
    return novedad


def cambiar_tecnico_novedad(
    db: Session,
    *,
    id_novedad: int,
    id_nuevo_tecnico: int,
    id_admin: int,
    motivo: Optional[str] = None,
) -> Novedad:
    """Cambia el técnico asignado al pedido/cita de una novedad."""
    novedad = db.query(Novedad).filter(Novedad.id_novedad == id_novedad).first()
    if not novedad:
        raise ValueError("Novedad no encontrada")

    tecnico_anterior_id = novedad.id_tecnico_n
    novedad.id_tecnico_n = id_nuevo_tecnico

    detalle_parts = [f"Técnico anterior: {tecnico_anterior_id}", f"Nuevo técnico: {id_nuevo_tecnico}"]
    if motivo:
        detalle_parts.append(f"Motivo: {motivo}")

    db.add(NovedadHistorial(
        id_novedad=id_novedad,
        id_usuario=id_admin,
        accion="Cambio de técnico",
        detalle=" | ".join(detalle_parts),
    ))
    db.commit()
    db.refresh(novedad)
    return novedad


# ── Cupones ─────────────────────────────────────────────────────────

def crear_cupon(
    db: Session,
    *,
    codigo: str,
    tipo_descuento: str,
    valor_descuento: float,
    id_cliente: int,
    id_novedad: Optional[int] = None,
    id_admin: int,
    fecha_vencimiento: Optional[date] = None,
    compra_minima: float = 0,
    usos_maximos: int = 1,
    aplica_tienda_completa: bool = True,
    categorias_aplicables: Optional[str] = None,
) -> CuponDescuento:
    """Crea un cupón de descuento asociado a una novedad y cliente."""
    # Verificar que el código no exista
    existente = db.query(CuponDescuento).filter(CuponDescuento.codigo == codigo).first()
    if existente:
        raise ValueError(f"Ya existe un cupón con el código {codigo}")

    cupon = CuponDescuento(
        codigo=codigo.upper(),
        tipo_descuento=tipo_descuento,
        valor_descuento=valor_descuento,
        fecha_vencimiento=fecha_vencimiento,
        compra_minima=compra_minima,
        usos_maximos=usos_maximos,
        aplica_tienda_completa=aplica_tienda_completa,
        categorias_aplicables=categorias_aplicables,
        id_cliente=id_cliente,
        id_novedad=id_novedad,
        id_admin_crea=id_admin,
        activo=True,
    )
    db.add(cupon)

    # Registrar en historial de novedad si existe
    if id_novedad:
        db.add(NovedadHistorial(
            id_novedad=id_novedad,
            id_usuario=id_admin,
            accion="Cupón generado",
            detalle=f"Cupón {codigo.upper()} — {tipo_descuento} {valor_descuento}%",
        ))

    db.commit()
    db.refresh(cupon)
    return cupon


def listar_cupones_cliente(db: Session, id_cliente: int) -> list[CuponDescuento]:
    """Lista cupones activos de un cliente."""
    return (
        db.query(CuponDescuento)
        .filter(
            CuponDescuento.id_cliente == id_cliente,
            CuponDescuento.activo == True,  # noqa: E712
        )
        .order_by(CuponDescuento.created_at.desc())
        .all()
    )


def validar_cupon(db: Session, codigo: str, id_cliente: int) -> Optional[CuponDescuento]:
    """Valida si un cupón es válido para un cliente. Retorna el cupón o None."""
    cupon = (
        db.query(CuponDescuento)
        .filter(
            CuponDescuento.codigo == codigo.upper(),
            CuponDescuento.activo == True,  # noqa: E712
        )
        .first()
    )
    if not cupon:
        return None
    if cupon.id_cliente and cupon.id_cliente != id_cliente:
        return None
    if cupon.fecha_vencimiento and cupon.fecha_vencimiento < date.today():
        return None
    if cupon.usos_realizados >= cupon.usos_maximos:
        return None
    return cupon

File: movil/services/novedades.service.ts
// ─────────────────────────────────────────────────────────────
// Servicio de novedades e incidencias para técnicos.
// Endpoints reales del backend (prefijo /api/v1/novedades):
// - GET  /novedades/tipos              → lista de tipos
// - GET  /novedades/estados            → lista de estados
// - GET  /novedades/prioridades        → lista de prioridades
// - GET  /novedades/clientes-buscar    → buscar clientes del técnico
// - GET  /novedades/clientes/:id/pedidos → pedidos del cliente para el técnico
// - GET  /novedades/clientes/:id/citas   → citas del cliente para el técnico
// - POST /novedades                    → crear novedad
// - POST /novedades/:id/evidencia      → subir evidencia (multipart)
// - GET  /novedades/mis-novedades      → listar novedades del técnico
// - GET  /novedades/:id                → detalle de novedad
// - GET  /novedades/:id/evidencias     → evidencias de una novedad
// ─────────────────────────────────────────────────────────────

import { apiFetch } from "./api";
import { API_BASE_URL } from "@/constants/api";
import { obtenerSesion } from "./storage";

// ── Interfaces ───────────────────────────────────────────────

export interface ClienteBusqueda {
  id_cliente: number;
  nombre: string;
  email: string | null;
  telefono: number | null;
  documento: number | null;
  direccion: string | null;
}

export interface PedidoCliente {
  id_pedido: number;
  fecha_pedido: string | null;
  estado_pedido: string | null;
  total: number | null;
  fecha_entrega: string | null;
  hora_entrega: string | null;
  hora_entrega_fin: string | null;
  estado_entrega: string | null;
  nombre_tecnico: string | null;
  detalles: { producto: string | null; cantidad: number | null; precio: number | null }[];
}

export interface CitaCliente {
  id_cita: number;
  fecha: string | null;
  hora: string;
  tipo_servicio: string;
  especialidad: string | null;
  estado: string;
  direccion: string;
  descripcion: string | null;
  costo: number | null;
  nombre_tecnico: string | null;
  id_tecnico: number | null;
}

export interface Novedad {
  id_novedad: number;
  tipo_novedad: string;
  descripcion_novedad: string;
  prioridad: string;
  estado_novedad: string;
  fecha_reporte: string | null;
  lugar_ocurrencia: string | null;
  evidencia_url: string | null;
  id_pedido: number | null;
  id_cita: number | null;
  id_cliente: number | null;
  id_devolucion: number | null;
  id_tecnico: number | null;
  tecnico_nombre: string | null;
  cliente_nombre: string | null;
  accion_admin: string | null;
  fecha_resolucion: string | null;
  evidencias: EvidenciaNovedad[];
  pedido_info: { id_pedido: number; estado: string | null; fecha_entrega: string | null } | null;
  cita_info: { id_cita: number; tipo_servicio: string; fecha: string | null; hora: string; estado: string } | null;
  devolucion_info: { id_devolucion: number; estado: string | null; motivo: string | null } | null;
}

export interface NovedadDetalle extends Novedad {
  historial: NovedadHistorial[];
  cliente_info: { id_cliente: number; nombre: string; email: string; telefono: number | null; documento: number | null; direccion: string | null } | null;
  pedido_info_completo: { id_pedido: number; estado: string | null; fecha_pedido: string | null; total: number | null; fecha_entrega: string | null; hora_entrega: string | null; estado_entrega: string | null; nombre_tecnico: string | null } | null;
  cita_info_completo: { id_cita: number; tipo_servicio: string; fecha: string | null; hora: string; estado: string; direccion: string; especialidad: string | null } | null;
  devolucion_info_completo: { id_devolucion: number; estado: string | null; motivo: string | null; descripcion: string | null; fecha_solicitud: string | null } | null;
}

export interface EvidenciaNovedad {
  id_evidencia_n: number;
  url: string;
  descripcion: string | null;
  fecha_subida: string | null;
}

export interface NovedadHistorial {
  id_historial: number;
  accion: string;
  detalle: string | null;
  fecha: string | null;
  usuario_nombre: string | null;
}

export interface NovedadCrearPayload {
  tipo_novedad: string;
  descripcion: string;
  prioridad?: string;
  id_pedido?: number | null;
  id_cita?: number | null;
  id_cliente?: number | null;
  id_devolucion?: number | null;
  lugar_ocurrencia?: string | null;
}

// ── Constantes ───────────────────────────────────────────────

export const TIPOS_NOVEDAD = [
  "Retraso en entrega",
  "Cliente ausente",
  "Cliente no recibió el pedido",
  "Dirección incorrecta",
  "Producto dañado",
  "Producto faltante",
  "Producto equivocado",
  "Pérdida de producto",
  "Robo de productos",
  "Accidente",
  "Problema con el vehículo/transporte",
  "Problema técnico",
  "Problema durante una cita",
  "Cita no realizada",
  "Cita reprogramada",
  "Cliente solicita reprogramación",
  "Problema con instalación",
  "Retraso que afecta cita",
  "Otro",
];

export const TIPOS_DEVOLUCION = [
  "Producto recibido con daños",
  "Producto incompleto",
  "Cliente no tenía el producto disponible",
  "Producto diferente al solicitado",
  "Devolución no realizada",
  "Problema durante la recogida",
  "Otro",
];

export const ESTADOS_NOVEDAD = [
  "Pendiente",
  "En revisión",
  "Aprobada",
  "Rechazada",
  "Resuelta",
  "Cerrada",
];

export const PRIORIDADES = ["baja", "normal", "alta", "urgente"];

// ── Funciones API ────────────────────────────────────────────

export function obtenerTiposNovedad() {
  return apiFetch<string[]>("/novedades/tipos");
}

export function obtenerEstadosNovedad() {
  return apiFetch<string[]>("/novedades/estados");
}

export function obtenerPrioridadesNovedad() {
  return apiFetch<string[]>("/novedades/prioridades");
}

export function buscarClientes(q?: string) {
  const params = q ? `?q=${encodeURIComponent(q)}` : "";
  return apiFetch<ClienteBusqueda[]>(`/novedades/clientes-buscar${params}`);
}

export function pedidosCliente(clienteId: number) {
  return apiFetch<PedidoCliente[]>(`/novedades/clientes/${clienteId}/pedidos`);
}

export function citasCliente(clienteId: number) {
  return apiFetch<CitaCliente[]>(`/novedades/clientes/${clienteId}/citas`);
}

export function crearNovedad(data: NovedadCrearPayload) {
  return apiFetch<{ id_novedad: number; mensaje: string }>("/novedades", {
    method: "POST",
    body: JSON.stringify(data),
  });
}

export function listarMisNovedades() {
  return apiFetch<Novedad[]>("/novedades/mis-novedades");
}

export function obtenerNovedad(id: number) {
  return apiFetch<NovedadDetalle>(`/novedades/${id}`);
}

export function obtenerEvidencias(novedadId: number) {
  return apiFetch<EvidenciaNovedad[]>(`/novedades/${novedadId}/evidencias`);
}

/**
 * Sube una evidencia (foto) a una novedad.
 * Usa multipart/form-data directamente con fetch (no apiFetch)
 * porque apiFetch no soporta FormData.
 */
export async function subirEvidencia(
  novedadId: number,
  archivo: { uri: string; name: string; type: string },
  descripcion?: string,
): Promise<{ id_evidencia_n: number; url: string; mensaje: string }> {
  const sesion = await obtenerSesion();
  const formData = new FormData();
  formData.append("file", archivo as any);
  if (descripcion) {
    formData.append("descripcion", descripcion);
  }

  const response = await fetch(
    `${API_BASE_URL}/novedades/${novedadId}/evidencia`,
    {
      method: "POST",
      headers: {
        Authorization: `Bearer ${sesion?.accessToken}`,
      },
      body: formData,
    },
  );

  if (!response.ok) {
    const err = await response.json().catch(() => ({ detail: "Error al subir evidencia" }));
    throw new Error(err.detail || "Error al subir evidencia");
  }

  return response.json();
}
